Core/clim-core/redefinition.lisp

```diff
@@ -0,0 +1,177 @@
+;;;; (C) Copyright 2019, 2020 Jan Moringen
+;;;;
+;;;; This library is free software; you can redistribute it and/or
+;;;; modify it under the terms of the GNU Library General Public
+;;;; License as published by the Free Software Foundation; either
+;;;; version 2 of the License, or (at your option) any later version.
+;;;;
+;;;; This library is distributed in the hope that it will be useful,
+;;;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+;;;; Library General Public License for more details.
+;;;;
+;;;; You should have received a copy of the GNU Library General Public
+;;;; License along with this library; if not, write to the
+;;;; Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+;;;; Boston, MA  02111-1307  USA.
+
+(in-package #:clim-internals)
+
+;;; Frame class redefinition notification protocol
+;;;
+;;; This is used for notifying application frame classes after aspects
+;;; such as the class itself, the `generate-panes' method or the
+;;; command table have been redefined. The notification sequence
+;;; begins with a call to `note-frame-class-redefined' which either
+;;; directly calls `handle-frame-class-redefinition' or schedules the
+;;; call to be made when leaving a
+;;; `with-delayed-redefinition-notifications'.
+;;;
+;;; This protocol is probably most commonly initiated by re-evaluating
+;;; the `define-application-frame' form that defined the frame class
+;;; in question, but any event that causes the `reinitialize-instance'
+;;; method of the class to execute also triggers it. In the context of
+;;; `define-application-frame',
+;;; `with-delayed-redefinition-notifications' delays the
+;;; `note-frame-class-redefinition' call until after all aspects
+;;; (`generate-panes' method, command table definition, etc) have been
+;;; executed.
+
+(defgeneric handle-frame-class-redefinition (class)
+  (:method ((class t))))
+
+(defvar *delayed-redefinition-notifications*)
+
+(defun call-with-delayed-redefinition-notifications (thunk)
+  (if (boundp '*delayed-redefinition-notifications*)
+      (funcall thunk)
+      (let ((*delayed-redefinition-notifications* '()))
+        (multiple-value-prog1
+            (funcall thunk)
+          (map nil #'handle-frame-class-redefinition
```

Comment:
Based on not using the result, `(map nil …)` is equally appropriate.

If we establish the convention that `map{c,car,…}` should be used for lists (as opposed to sequences), `mapc` would make sense.

Reply:
mapc is more specific for two reasons:
1. it clearly shows that the collection is a list
2. it clearly shows that nothing depends on the result

map nil may be interpreted, that it works on any sequence and that the code invoking the form depends on it returning NIL

Reply:
> it clearly shows that the collection is a list

I can understand this and it is a good enough reason to make the change (assuming we do this everywhere as part of a consistent coding style).

> it clearly shows that nothing depends on the result […] and that the code invoking the form depends on it returning NIL

I don't understand this part. `mapc` is specified to return `list-1` and somebody could rely on that return value just as well.

But as I wrote above, I can make the change since the sequence is a list in this case.